opentasks-provider/src/main/java/org/dmfs/provider/tasks/processors/tasks/Instantiating.java

```diff
@@ -75,141 +87,143 @@ public Instantiating(EntityProcessor<TaskAdapter> delegate)
+                TaskDatabaseHelper.Tables.INSTANCE_VIEW,
+                new String[] { TaskContract.Instances._ID, TaskContract.Instances.INSTANCE_ORIGINAL_TIME },
+                String.format(Locale.ENGLISH, "%s = %d", TaskContract.Instances.TASK_ID, id),
+                null,
+                null,
+                null,
+                TaskContract.Instances.INSTANCE_ORIGINAL_TIME);
+
+
+        /*
+         * The goal of the code below is to update existing instances in place (as opposed to delete and recreate all instances). We do this for two reasons:
+         * 1) efficiency, in most cases existing instances don't change, deleting and recreating them would be very expensive
+         * 2) stable row ids, deleting and recreating instances would change their id and void any existing URIs to them
+         */
+        try
+        {
+            int idIdx = existingInstances.getColumnIndex(TaskContract.Instances._ID);
+            final int startIdx = existingInstances.getColumnIndex(TaskContract.Instances.INSTANCE_ORIGINAL_TIME);
+
+            // get an Iterator of all expected instances
+            // for very long or even infinite series we need to stop iterating at some point.
+            // TODO: once we actually support recurrence we should only count future instances
+
+            Iterable<Pair<Optional<ContentValues>, Optional<Integer>>> diff = new Diff<>(
+                    new Mapped<>(new SingleValueFunction<ContentValues>(), new Limited<>(INSTANCE_COUNT_LIMIT, new InstanceValuesIterable(taskAdapter))),
+                    new Range(existingInstances.getCount()),
+                    new BiFunction<ContentValues, Integer, Integer>()
+                    {
+                        @Override
+                        public Integer value(ContentValues newInstanceValues, Integer cursorRow)
+                        {
+                            existingInstances.moveToPosition(cursorRow);
+                            return (int) (existingInstances.getLong(startIdx) - newInstanceValues.getAsLong(TaskContract.Instances.INSTANCE_ORIGINAL_TIME));
+                        }
+                    });
+
+            // sync the instances table with the new instances
+            for (Pair<Optional<ContentValues>, Optional<Integer>> next : diff)
+            {
+                if (!next.left().isPresent())
+                {
+                    // there is no new instance for this old one, remove it
+                    existingInstances.moveToPosition(next.right().value());
+                    db.delete(TaskDatabaseHelper.Tables.INSTANCES,
+                            String.format(Locale.ENGLISH, "%s = %d", TaskContract.Instances._ID, existingInstances.getLong(idIdx)), null);
+                }
+                else if (!next.right().isPresent())
+                {
+                    // there is no old instance for this new one, add it
+                    ContentValues values = next.left().value();
```

Comment:
I wanted to check the coverage here in the IDE, to see if `TaskProviderTest` covers this, but since it's an android test, it is not possible of course. I've checked in a different project that Roboelectric tests can be run with coverage. So I am wondering whether it would be useful to add a Roboelectric version as well of this test.
Android test is still desirable and the more important one, I think, because of reasons mentioned earlier, but it wouldn't be a big effort to add the same test (with code-reuse) as a roboelectric one as well, to be able to see code coverage of these rather complex codes in the provider.
(It probably requires a shared source set between test and androidTest.)
What do you think?

Reply:
I'm wondering if there isn't a way to get the coverage of Android tests.
I wouldn't want to duplicate the tests for Robolectric.

Also, we have much more options to test all the provider classes in isolation using Robolectric. Most of them (at least the newer ones) should be quite testable.